### PR TITLE
fix(core): Use AbortController to notify nodes to abort execution

### PR DESCRIPTION
**Pull request by netroy**
and add support for cancelling ongoing operations inside a node.

### CODE REVIEW

File: packages/nodes-base/nodes/Wait/Wait.node.ts
Comment (netroy) [inline packages/nodes-base/nodes/Wait/Wait.node.ts:864]: This should be moved out of nodes code.

**Comment by netroy** [comment]
closing this for now. will try to come up with a simpler mechanism to abort executions at individual node level.

**Comment by OlegIvaniv** [comment]
@netroy I've been seeing a lot of `MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit` errors lately with these changes, any idea what might be wrong?

**Comment by netroy** [comment]
> @netroy I've been seeing a lot of `MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit` errors lately with these changes, any idea what might be wrong?

I'm guessing that this is coming from the fact that every node you pass down the `AbortSignal` into is going to add a listener on it. And since a workflow could have dozens (if not hundreds) of nodes using this mechanism, we need to do a `(abortSignal as unknown as EventEmitter).setMaxListeners(Infinity);` when the `AbortController` is created.

I'll fix that in this PR. but, since this is just a warning, I assume that your aren't blocked by this.

**Comment by OlegIvaniv** [comment]
> > @netroy I've been seeing a lot of `MaxListenersExceededWarning: Possible EventTarget memory leak detected. 11 abort listeners added to [AbortSignal]. Use events.setMaxListeners() to increase limit` errors lately with these changes, any idea what might be wrong?
> 
> I'm guessing that this is coming from the fact that every node you pass down the `AbortSignal` into is going to add a listener on it. And since a workflow could have dozens (if not hundreds) of nodes using this mechanism, we need to do a `(abortSignal as unknown as EventEmitter).setMaxListeners(Infinity);` when the `AbortController` is created.
> 
> I'll fix that in this PR. but, since this is just a warning, I assume that your aren't blocked by this.

Ah, understood. No, not blocked, was just worried it might have been something more nefarious. Sounds like we're all good then. Appreciate the context.

File: packages/core/src/WorkflowExecute.ts
Comment (netroy) [inline packages/core/src/WorkflowExecute.ts:822]: once this PR is merged, I'm going to get rid of `PCancelable` and use a regular promise with `AbortController` instead. 
Even `p-cancelable` [recommends](https://github.com/sindresorhus/p-cancelable/commit/5e8b26853d994c1de14232ff839945fc331d2912) to do that since 3 years.
Moving away from `PCancelable` should also make this code more readable.

File: packages/core/src/WorkflowExecute.ts
Comment (ivov) [inline packages/core/src/WorkflowExecute.ts:829]: Looks like we can pass in a `reason` arg - all cancellations that can happen right now are user-requested? Else we might want to use this arg to differentiate between user-requested cancellations vs. cancellations initiated on our end?

File: packages/editor-ui/src/mixins/pushConnection.ts
Comment (ivov) [inline packages/editor-ui/src/mixins/pushConnection.ts:295]: We're using `canceled` internally but `stopped` externally - we might want to unify for consistency.

File: packages/core/src/NodeExecuteFunctions.ts
Comment (ivov) [inline packages/core/src/NodeExecuteFunctions.ts:2519]: Would it be useful to log here that a cancellation occurred?

File: packages/core/src/NodeExecuteFunctions.ts
Comment (ivov) [inline packages/core/src/NodeExecuteFunctions.ts:2515]: Would it make sense to unit test this?

File: packages/cli/src/services/dynamicNodeParameters.service.ts
Comment (ivov) [inline packages/cli/src/services/dynamicNodeParameters.service.ts:110]: For understanding - how are we going to abort this if we don't keep a reference to the controller? Same in `CredentialsHelper`. Or is this only to satisfy the signature? (If the latter, maybe this should be optional?)

File: packages/cli/src/databases/repositories/execution.repository.ts
Comment (netroy) [reply packages/cli/src/databases/repositories/execution.repository.ts]: Added this because of a previous change, which has since been reverted. reverted this as well.

File: packages/cli/src/ActiveExecutions.ts
Comment (netroy) [reply packages/cli/src/ActiveExecutions.ts]: This was needed before I added the [`setTimeout` in WorkflowExecute](https://github.com/n8n-io/n8n/pull/6141/files#diff-0bfa6f1919bce33f0a8185d4496af9b38028eee2ed829ae01b225e2ff42a294bR832) because the executions was getting removed in the `.cancel` call.
reverted.

File: packages/core/src/WorkflowExecute.ts
Comment (netroy) [reply packages/core/src/WorkflowExecute.ts:829]: Good idea. we can have cancellation from user interaction, timeout, or server shutdown.
I've created a backlog ticket: [ENG-92](https://linear.app/n8n/issue/ENG-92).

File: packages/editor-ui/src/mixins/pushConnection.ts
Comment (netroy) [reply packages/editor-ui/src/mixins/pushConnection.ts:295]: Agreed that we should unify these. Created a ticket [ENG-93](https://linear.app/n8n/issue/ENG-93)

File: packages/core/src/NodeExecuteFunctions.ts
Comment (netroy) [reply packages/core/src/NodeExecuteFunctions.ts]: Do you mean logging for every node being aborted? or do you mean we should log when an execution is aborted.
This code is called for individual node that wishes to do something on execution cancellation.

File: packages/cli/src/services/dynamicNodeParameters.service.ts
Comment (netroy) [reply packages/cli/src/services/dynamicNodeParameters.service.ts]: reverted

File: packages/core/src/WorkflowExecute.ts
Comment (netroy) [reply packages/core/src/WorkflowExecute.ts:832]: yeah. we have a bunch of async operations (that don't really take any time, and possibly don't even need to be async) that need to happen between the cancellation, and before this `resolve` is called. I don't like this, and plan to fix this in the same PR where I'll remove `PCancelable` completely.

File: packages/core/src/NodeExecuteFunctions.ts
Comment (netroy) [reply packages/core/src/NodeExecuteFunctions.ts:2515]: not sure. I generally prefer to unit tests stuff that has logic in it. not sure if it make sense to unit test if the event listeners is correctly attached. maybe it makes sense to test that the handler is invoked only once, but I think that can happen later.

File: packages/core/src/NodeExecuteFunctions.ts
Comment (ivov) [reply packages/core/src/NodeExecuteFunctions.ts]: Exactly, I meant for every node aborted - but on second thought, it might be enough to log when the execution itself is cancelled.

**Review by ivov — DISMISSED** [review]
Thanks for explaining and addressing everything!

File: packages/core/src/NodeExecuteFunctions.ts
Comment (netroy) [reply packages/core/src/NodeExecuteFunctions.ts]: for that to work, we'd have to add the log to `onCancel`, while all the other execution finishing logs are in `processSuccessExecution`. After `PCancelable` is gone, this will be less mess, and I'll add this logging then.

**Review by OlegIvaniv — APPROVED** [review]

**Pull request merged**
